drivers:platform:mbed: GPIO platform driver implementation

mphalke opened the pull request:
Implemented the GPIO read/write functionality for Mbed platform.

Signed-off-by: mahphalke <Mahesh.Phalke@analog.com>

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+
+		/* Set the GPIO value */
+		if (gpio_set_value(desc, value) != SUCCESS) {
+			free(gpio_output);
```
Should we use `delete` to correctly pair with `new`?

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+		gpio_desc_extra->direction = GPIO_OUT;
+
+		/* Set the GPIO value */
+		if (gpio_set_value(desc, value) != SUCCESS) {
```
Would it be better to use `gpio_output->write(value)`? `gpio_set_value` seems a bit unclear what it will do.

mphalke replied:
This is good point. We can use write() function directly.

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+
+	if (gpio_desc_extra->gpio_pin) {
```
Is it worthwhile to get the direction first to make sure the pin is output? What happens if we try to write an input pin?

mphalke replied:
Valid point. Added check for GPIO direction.

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+
+	gpio_input = (DigitalIn *)((struct mbed_gpio_desc *)desc->extra)->gpio_pin;
+	*value = (uint8_t)gpio_input->read();
+	ret = gpio_input->is_connected() ? SUCCESS : -EIO;
```
Question: should this be done before trying to read the GPIO value? And should we do this in every function before this?

mphalke replied:
Good point. I think it doesn't make any harm to use this check for every function as it just checks if pin is mapped to valid GPIO and not NC.

amiclaus commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+
+	if (gpio_desc_extra->gpio_pin)
```
I think you should do something like:
```
if (!(gpio_desc_extra->gpio_pin))
    return -ENODEV (or -EINVAL);

*direction = gpio_desc_extra->direction;

return SUCCESS;
```

mphalke replied:
Make sense. Updated the code accordingly.

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+	if (gpio_desc_extra->gpio_pin == NULL) {
```
Suggestion: I think we could improve the code if we check `if (gpio_desc_extra->gpio_pin != NULL)` and do a `delete gpio_desc_extra->gpio_pin;` if we have instantiated it before. This can make the code easier to read, in my opinion, and can improve functionality by allowing us to change the direction of the pin after we set it once.

This change is not necessary if you think it doesn't make sense. Let me know.

adrimbarean replied:
May also apply to `mbed_gpio_direction_output`.

mphalke replied:
Agreed. I will update code to simplify it.

adrimbarean commented on drivers/platform/mbed/mbed_gpio.cpp:
```diff
@@ -0,0 +1,308 @@
+
+	/* Free the gpio object */
+	if (((struct mbed_gpio_desc *)(desc->extra))->gpio_pin)
+		free(((struct mbed_gpio_desc *)(desc->extra))->gpio_pin);
```
I believe this is allocated with `new` always so we should use `delete` to clean it up. It will have undefined behavior otherwise.

mphalke replied:
gpio_pin can be either digital_in Or digital_out instance. Use of delete needs specific typecast to the object and compiler gives warning if not typecasted, so using free here. Is this fine to use free or should we use delete anyway?

buha replied:
I also think this should be done with `delete`, it's the proper way to free resources allocated with `new` (performs something more than `free`)

adrimbarean reviewed (commented):
A few questions and comments for now.

mphalke commented:
Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?

amiclaus reviewed (commented):
> Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?

We should not have commits like `Fix review comments` in the history of our master branch.
Therefore I'd suggest squashing them

mphalke commented:
> > Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?
> 
> We should not have commits like `Fix review comments` in the history of our master branch. Therefore I'd suggest squashing them

I got it, thanks. I have squashed the commits and forced pushed.

amiclaus reviewed (commented):
> > > Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?
> > 
> > 
> > We should not have commits like `Fix review comments` in the history of our master branch. Therefore I'd suggest squashing them
> 
> I got it, thanks. I have squashed the commits and forced pushed.

But we still need to keep track of the changes that were made based on the review. Therefore a changelog should be added. See example: https://github.com/analogdevicesinc/no-OS/pull/1240#issuecomment-1029845317

mphalke commented:
> > > > Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?
> > > 
> > > 
> > > We should not have commits like `Fix review comments` in the history of our master branch. Therefore I'd suggest squashing them
> > 
> > 
> > I got it, thanks. I have squashed the commits and forced pushed.
> 
> But we still need to keep track of the changes that were made based on the review. Therefore a changelog should be added. See example: [#1240 (comment)](https://github.com/analogdevicesinc/no-OS/pull/1240#issuecomment-1029845317)

How do I generate the change logs as I have already squashed them. Any suggestions?

amiclaus reviewed (commented):
> > > > > Hi, I have pushed the changes. Do you want me to squash the last 2 commits Or two separate commits are okay?
> > > > 
> > > > 
> > > > We should not have commits like `Fix review comments` in the history of our master branch. Therefore I'd suggest squashing them
> > > 
> > > 
> > > I got it, thanks. I have squashed the commits and forced pushed.
> > 
> > 
> > But we still need to keep track of the changes that were made based on the review. Therefore a changelog should be added. See example: [#1240 (comment)](https://github.com/analogdevicesinc/no-OS/pull/1240#issuecomment-1029845317)
> 
> How do I generate the change logs as I have already squashed them. Any suggestions?

You don't generate them. You mention what changed after your last force push.

mphalke commented:
V2:

- Improved the code readability by restructuring the negative error check. Negative error checks are added at the start of all functions.
- Added GPIO mapping validation using 'is_connected()' function in all the GPIO read/write functions.
- Added check for GPIO direction before writing to input GPIO.

mphalke commented:
V3: 
- Using 'delete' to free the memory allocated for mbed GPIO descriptor instead of 'free' in gpio_remove() function.
